### ISSUE TITLE
How to make socket.io work well with pm2 when start more than one instance?

### ISSUE THREAD

**Issue by node-rookie** [Unitech/PM2#1510]

**Comment by klinquist** [comment]
In cluster mode, things should work fine.  In fork mode, you'd need to run your socket.io server on a different port for each instance and reverse proxy to them via nginx (with ip_hash set) or HAProxy...

**Comment by thewilli** [comment]
@klinquist I have the same question as the OP. socket.io requires that each client always connects to the same worker, but it doesn't seem that the pm2 loadbalancer does some internal magic to ensure it, right? So the only option would be to use some external load balancer (like nginx) and fixes ports. But how can I get pm2 to give each worker a fixed port number that persists even a crash and restart of a worker?

**Comment by klinquist** [comment]
@thewilli   

Two things:
1)    "socket.io requires that each client always connects to the same worker"  -  says who?
A properly written nodejs application should be stateless!  Clients should absolutely not have to connect to the same worker.  

2)  You could map ports to the NODE_INSTANCE environment variable (which I believe PM2 sets ...   that name might not be right.....  check process.env.NODE_ENV  to verify).  But once again, doing it this way is bad practice!!

**Comment by jshkurti** [comment]
@thewilli When you run an app in cluster mode with PM2 each worker will receive a value in `process.env.NODE_APP_INSTANCE` which goes from 0 to number of workers-1.
Let's say you start your app with `pm2 start app.js -i 4`, then your worker will receive 0 1 2 3 as NODE_APP_INSTANCE value.

Then in your code you can use that value to set the listen port such as 
`server.listen(8000 + process.env.NODE_APP_INSTANCE)`

**Comment by thewilli** [comment]
@klinquist 
the [docs](http://socket.io/docs/using-multiple-nodes/) say it, and I tried it by myself:

> If you plan to distribute the load of connections among different processes or machines, you have to make sure that requests associated with a particular session id connect to the process that originated them.

**Comment by thewilli** [comment]
@jshkurti works great, thank you!

**Comment by jiajianrong** [comment]
Just wondering if the only solution for socket.io is to use [sticky session](https://github.com/indutny/sticky-session)?

**Comment by karlpokus** [comment]
@jiajianrong I believe `long-polling` is the main culprit here and can be toggled. Instructions [here](https://socket.io/docs/client-api/#with-websocket-transport-only)

**Comment by mattbrunetti** [comment]
Thank you for your note @karlpokus .. I was under the impression that WebSockets themselves required a stateful server, but of course I was wrong

TL;DR

PM2 works fine with socket.io if you [use 'websocket' transport only](https://socket.io/docs/client-api/#with-websocket-transport-only), which does not require a [stateful server](http://pm2.keymetrics.io/docs/usage/cluster-mode/#statelessify-your-application). Other transports (i.e. 'polling') require a stateful server. WebSockets are compatible with IE 11 a

**Comment by qiulang** [comment]
I can very that except for the latest firefox (56), chrome/safari/opera all work when using 'websock' only with pm2 cluster.

Firefox can’t establish a connection to the server at ws://localhost:3000/socket.io/?EIO=3&transport=websocket.

The connection to ws://localhost:3000/socket.io/?EIO=3&transport=websocket was interrupted while the page was loading.

**Comment by knoxcard** [comment]
thank you @jshkurti!  I couldn't find your solution anywhere else on the web.

Here is what my end code looks like...

**SERVER**
```
  var http = require('http')
  http.globalAgent.maxSockets = Infinity
  app.server = http.Server(app)
  app.server.listen(parseInt(app.nconf.get('app:port')) + (process.env.NODE_APP_INSTANCE ? parseInt(process.env.NODE_APP_INSTANCE) : 0), () => {
    app.sio = require('socket.io').listen(app.server, {
      transports: [
        'websocket',
        'polling',
        'long-polling'
      ]
    })
    app.sio.adapter(require('socket.io-redis')({
      host: 'localhost',
      port: 6379
    }))
    app.sio.use((socket, next) => {
      app.sessionMiddleware(socket.request, socket.request.res, next)
    })
    app.sio.sockets.on('connection', socket => {
      var room = 'room_' + (socket.request.session.user_id ? socket.request.session.user_id : socket.request.session.id)
      socket.join(room)
      app.sio.sockets.in(room).emit('auth', {
        action: 'check'
      })
    })
    app.sio.sockets.on('disconnect', () => {
      console.log(app.colors.red('[socket disconnected]'))
    })
  })
```
**CLIENT**
```
	jQuery.socket = io()
	jQuery.socket.on('connect', function() {
		jQuery.socket.on('auth', function(ret) {
		})
	})
	jQuery.socket.on('disconnect', function() {
		jQuery.reconnect = null
		jQuery.reconnect = setInterval(function() {
			jQuery.socket = io.connect()
			console.log('Attempting server connection...')
		}, 10000)
	})
```

**Comment by subatomicglue** [comment]
I see no mention of redis.

Redis pub/sub should coordinate a shared state between all the pm2 workers... so shouldn't that allow polling to work?

**Comment by tvvignesh** [comment]
@jshkurti If we use the instance env to generate one port for every node like what you said, how would we expose the same via a reverse proxy like nginx? What port number should we specify there?

**Comment by Gby56** [comment]
@tvvignesh yeah you loose the loadbalancing from pm2 if you use @jshkurti 's solution, and end up doing it with nginx 😅

**Comment by hdodov** [comment]
> @thewilli When you run an app in cluster mode with PM2 each worker will receive a value in `process.env.NODE_APP_INSTANCE` which goes from 0 to number of workers-1.
> Let's say you start your app with `pm2 start app.js -i 4`, then your workers will receive 0 1 2 3 as NODE_APP_INSTANCE value.
> 
> Then in your code you can use that value to set the listen port such as
> `server.listen(8000 + process.env.NODE_APP_INSTANCE)`

@jshkurti thanks for the insight! I don't get one thing. My workers will listen on the respective ports (in this case, 8001, 8002...). However, I use nginx as a reverse proxy and how should I configure it to forward the requests to these ports? I mean, I forward `/socket.io` requests to `http://localhost:8000/` and if I keep doing that, I'll still be using only one worker. The others would sit around and do nothing, right?

**Comment by PsyGik** [comment]
@hdodov, Maybe this way

````
http {
    upstream myapp1 {
        server srv1.example.com:8001;
        server srv1.example.com:8002;
        server srv1.example.com:8003;
    }

    server {
        listen 80;
        location / {
            proxy_pass http://myapp1;
        }
    }
}
````

**Comment by averri** [comment]
@jshkurti, what you have mentioned about changing the port numbers for pm2 **cluster** mode is not required. pm2 manages to share the same port across different process - Reference: https://pm2.keymetrics.io/docs/usage/cluster-mode.

If you are changing the ports of your application for using the pm2 cluster mode you are doing it wrong!
